src/docfx/serve/lsp/TextDocumentHandler.cs

```diff
@@ -37,31 +38,39 @@ internal class TextDocumentHandler : ITextDocumentSyncHandler
 
         public Task<Unit> Handle(DidChangeTextDocumentParams notification, CancellationToken token)
         {
+            var eventTimeStamp = DateTime.UtcNow;
             if (!TryUpdatePackage(notification.TextDocument.Uri, notification.ContentChanges.First().Text))
             {
                 _notificationListener.OnNotificationHandled();
                 return Unit.Task;
             }
 
-            _languageServerBuilder.QueueBuild();
+            _languageServerBuilder.QueueBuild(eventTimeStamp);
             return Unit.Task;
         }
 
         public Task<Unit> Handle(DidOpenTextDocumentParams notification, CancellationToken token)
         {
+            var eventTimeStamp = DateTime.UtcNow;
             if (!TryUpdatePackage(notification.TextDocument.Uri, notification.TextDocument.Text))
             {
                 _notificationListener.OnNotificationHandled();
                 return Unit.Task;
             }
 
-            _languageServerBuilder.QueueBuild();
+            _languageServerBuilder.QueueBuild(eventTimeStamp);
             return Unit.Task;
         }
 
         public Task<Unit> Handle(DidCloseTextDocumentParams notification, CancellationToken token)
         {
-            _notificationListener.OnNotificationHandled();
+            var eventTimeStamp = DateTime.UtcNow;
```

Comment:
We may not need this timestamping according to https://github.com/OmniSharp/csharp-language-server-protocol/issues/249#issuecomment-636491490

> With the current implementation you should never see concurrent threads for "write" (Serial) operations (didchange, rename, etc). You could however have multiple threads "reading" (Parallel) at the same time (for completions and such).